radix-engine/src/engine/errors.rs

```diff
@@ -161,25 +206,76 @@ pub enum ModuleError {
     ExecutionTraceError(ExecutionTraceError),
 }
 
-impl Into<ModuleError> for AuthError {
-    fn into(self) -> ModuleError {
-        ModuleError::AuthError(self)
+impl CanBeAbortion for ModuleError {
+    fn abortion(&self) -> Option<&AbortReason> {
+        match self {
+            Self::CostingError(err) => err.abortion(),
+            _ => None,
+        }
+    }
+}
+
+impl From<NodeMoveError> for ModuleError {
+    fn from(error: NodeMoveError) -> Self {
+        Self::NodeMoveError(error)
+    }
+}
+
+impl From<AuthError> for ModuleError {
+    fn from(error: AuthError) -> Self {
+        Self::AuthError(error)
+    }
+}
+
+impl From<CostingError> for ModuleError {
+    fn from(error: CostingError) -> Self {
+        Self::CostingError(error)
+    }
+}
+
+impl From<RoyaltyError> for ModuleError {
+    fn from(error: RoyaltyError) -> Self {
+        Self::RoyaltyError(error)
+    }
+}
+
+impl From<ExecutionTraceError> for ModuleError {
+    fn from(error: ExecutionTraceError) -> Self {
+        Self::ExecutionTraceError(error)
     }
 }
 
 #[derive(Debug)]
-pub enum InvokeError<E> {
+pub enum InvokeError<E: InvokeWrappedError> {
     Error(E),
     Downstream(RuntimeError),
 }
 
-impl<E> From<RuntimeError> for InvokeError<E> {
+/// This is a trait for the non-Downstream part of [`InvokeError`]
+/// We can't use `Into<RuntimeError>` because we need `RuntimeError` _not_ to implement it.
+pub trait InvokeWrappedError {
```

Comment:
How about renaming to `NonDownstreamError` or `SelfError`?

Reply:
Sounds good! I struggled with finding a good name